pkg/opendsme/contrib/DSMEPlatform.cpp

```diff
@@ -0,0 +1,738 @@
+
+    releaseMessage(m);
+}
+
+void DSMEPlatform::handleDataMessageFromMCPS(DSMEMessage *msg)
+{
+    msg->dispatchMessage();
+}
+
+bool DSMEPlatform::isReceptionFromAckLayerPossible()
+{
+    return this->state == STATE_READY;
+}
+
+void DSMEPlatform::handleReceivedMessageFromAckLayer(IDSMEMessage *message)
+{
+    DSME_ASSERT(receiveFromAckLayerDelegate);
+    DSME_ASSERT(!this->message);
+    this->message = message;
+    event_post(this->getEventQueue(), &rx_offload_ev);
+}
+
+DSMEMessage *DSMEPlatform::getEmptyMessage()
+{
+    DSMEMessage *msg = new DSMEMessage();
+
+    DSME_ASSERT(msg);
+    msg->clearMessage();
+    signalNewMsg(msg);
+    return msg;
+}
+
+void DSMEPlatform::signalNewMsg(DSMEMessage *msg)
+{
+    /* Not used */
+}
+
+void DSMEPlatform::releaseMessage(IDSMEMessage *msg)
+{
+    DSMEMessage *m = static_cast<DSMEMessage *>(msg);
+
+    m->releaseMessage();
+}
+
+void DSMEPlatform::startTimer(uint32_t symbolCounterValue)
+{
+    uint32_t now = ztimer_now(ZTIMER_USEC);
+    uint32_t offset = now & OPENDSME_TIMER_MASK;
+    /* This works even if there's an overflow */
+    int32_t delta = ((symbolCounterValue - getSymbolCounter()) << OPENDSME_TIMER_OFFSET)
+                    - offset;
+
+    ztimer_set(ZTIMER_USEC, &timer, (uint32_t)delta);
+}
+
+uint32_t DSMEPlatform::getSymbolCounter()
+{
+    return ztimer_now(ZTIMER_USEC) >> OPENDSME_TIMER_OFFSET;
+}
+
+void DSMEPlatform::scheduleStartOfCFP()
+{
+    event_post(this->getEventQueue(), &start_of_cfp_ev);
+}
+
+void DSMEPlatform::signalAckedTransmissionResult(bool success, uint8_t transmissionAttempts,
+                                                 IEEE802154MacAddress receiver)
+{
+    /* Not used */
+}
+
+/*
+ * Signal GTS allocation or deallocation
+ */
+void DSMEPlatform::signalGTSChange(bool deallocation, IEEE802154MacAddress counterpart,
+                                   uint16_t superframeID, uint8_t gtSlotID, uint8_t channel,
+                                   Direction direction)
+{}
+
+void DSMEPlatform::signalQueueLength(uint32_t length)
+{}
+
+/*
+ * Number of packets sent per CAP
+ */
+void DSMEPlatform::signalPacketsPerCAP(uint32_t packets)
+{}
+
+/*
+ * Number of failed packets per CAP
+ */
+void DSMEPlatform::signalFailedPacketsPerCAP(uint32_t packets)
+{
+    /* Not used */
+}
+
+void DSMEPlatform::updateVisual()
+{
+    /* Not used */
+}
+
+bool DSMEPlatform::setChannelNumber(uint8_t channel)
+{
+    ieee802154_phy_conf_t conf = {
+        .phy_mode = IEEE802154_PHY_OQPSK,
+        .channel = channel,
+        .page = 0,
+        .pow = CONFIG_IEEE802154_DEFAULT_TXPOWER,
+    };
+    int res;
+
+    res = ieee802154_radio_set_idle(this->radio, true);
+    DSME_ASSERT(res == 0);
+    res = ieee802154_radio_config_phy(this->radio, &conf);
+    DSME_ASSERT(res == 0);
+
+    /* TODO: Find a better solution */
+    ieee802154_radio_config_addr_filter(this->radio, IEEE802154_AF_PANID, &this->mac_pib.macPANId);
```

Comment:
this will set the panID every time the channel changes. It's only because the MAC updates the PAN ID on association and does not inform the platform layer about the change. Therefore, with that snippet I make sure it's always right.